src/KubernetesClient/Controllers/ResourceEventDeltaBlock.cs

```diff
@@ -0,0 +1,284 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reactive.Disposables;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using k8s.Informers.Notifications;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+namespace k8s.Controllers
+{
+    /// <summary>
+    ///     A TPL Dataflow block exposes queuing semantics that attaches to infromer observable streams.
+    ///     It groups delta changes per key, sending each group to downstream blocks as individual messages
+    ///     This allows "Per resource ID" processing semantics
+    ///     Any sync events are only propagated if no other changes are queued up
+    /// </summary>
+    /// <typeparam name="TKey">The type of key used to identity resource</typeparam>
+    /// <typeparam name="TResource">The resource type</typeparam>
+    public class ResourceEventDeltaBlock<TKey, TResource> : ISourceBlock<List<ResourceEvent<TResource>>>, ITargetBlock<ResourceEvent<TResource>>
+    {
+        private readonly Dictionary<TKey, List<ResourceEvent<TResource>>> _items = new Dictionary<TKey, List<ResourceEvent<TResource>>>();
+        private readonly Func<TResource, TKey> _keyFunc;
+        /// <summary>
+        /// Protects access to _queue, _targets and _items as well as ensuring some operations are executed in atomic way
+        /// </summary>
+        private readonly object _lock = new object();
+        private readonly ILogger _log;
+        private readonly Queue<List<ResourceEvent<TResource>>> _queue = new Queue<List<ResourceEvent<TResource>>>();
+        private readonly bool _skipTransient;
+        private readonly List<TargetLink> _targets = new List<TargetLink>();
+        private readonly TaskCompletionSource<object> _taskCompletionSource = new TaskCompletionSource<object>();
+        private bool _isCompleting;
+        private long _msgId;
+
+        /// <param name="keyFunc">The key selector function</param>
+        /// <param name="transient"></param>
+        /// <param name="skipTransient">If <see langword="true">, removes any batches in which resource was created (first msg=Add) and removed (last msg=Delete)</param>
+        public ResourceEventDeltaBlock(Func<TResource, TKey> keyFunc, bool skipTransient = true, ILogger log = null)
+        {
+            log ??= NullLogger.Instance;
+            _keyFunc = keyFunc;
+            _skipTransient = skipTransient;
+            _log = log;
+        }
+
+        public void Complete()
+        {
+            _isCompleting = true;
+            SetCompletedIfNeeded();
+        }
+
+        public void Fault(Exception exception)
+        {
+            _taskCompletionSource.SetException(exception);
+            foreach (var link in _targets.Where(x => x.LinkOptions.PropagateCompletion)) link.Target.Fault(exception);
```

Comment:
The enumeration must be done with the lock held; otherwise, it's possible an `InvalidOperationException` will be thrown if the collection is mutated concurrently.